### ISSUE TITLE
Add ledger hardware wallet support

### ISSUE THREAD

**Issue by danfinlay** [MetaMask/metamask-extension#717]
Would be cool to support this hardware wallet:
https://www.ledgerwallet.com

Currently blocked by #3249

**Comment by flyswatter** [comment]
Most of the JS integration has already been written by the MyEtherWallet team [here](https://github.com/kvhnuke/etherwallet/tree/mercury/app/scripts/staticJS)!

**Comment by izqui** [comment]
This would be really cool indeed! Happy to help make it possible

Now that this milestone is done https://github.com/MetaMask/metamask-plugin/milestone/14?closed=1, would it be possible to implement Ledger support?

**Comment by flyswatter** [comment]
Hi @izqui, thanks for the interest!

Our architecture will now support it, but we still haven't settled on what the UI/UX is going to be.  If you know the constraints of the Ledger well, maybe you could help propose an account adding / signing flow.

One proposal:

1.  We're planning to add an "Import Account" menu item to the top-right menu.
2. That menu will allow selection of different types of accounts to support.
3. One of those account types could be Ledger.

What does that UI look like? What code can we use under the hood?

The other big question for Ledger support:

What does transaction signing look like?  I generally like the idea that reviewing a transaction can continue to have a common interface, and as we improve that, it improves all signing strategies, but when the user clicks "Accept", some types of accounts (like Ledger) will need to show another view, and instruct the user how to complete the interaction.  What does *that* look like?

**Comment by kumavis** [comment]
@izqui also - how does a ledger wallet integration work? do you start a webserver stored on the USB device? Haven't looked into it yet but certainly is possible.

**Comment by flyswatter** [comment]
Last time I asked, @kumavis, I was told to basically copy MEW's code, but it didn't look quite as neatly abstracted as I'd hoped.  I'm hoping there'd be a Ledger.js lib that we could drop in for getting listed accounts & requesting signatures.

**Comment by izqui** [comment]
@flyswatter My Ledger Nano S is arriving on Friday. Will think about what a good flow could be then. 

Regarding signing transactions, i think showing the Metamask confirmation screen but without buttons, indicating to perform the confirmation using Ledger's hardware could be an option.

**Comment by izqui** [comment]
I have been doing some research and it seems that the Ledger only allows to sign transactions and not arbitrary data.

AFAIK, this would make deploying contracts or signing proofs impossible.
 
Also, regarding My Ether Wallet implementation:
* MEW doesn't allow using Ledger for signing messages (https://www.myetherwallet.com/#sign-message) 
* MEW offers the option to deploy a contract using the Ledger, but I cannot get it to work (https://www.myetherwallet.com/#contracts)

I have talked to Ledger support to see if they can give me an status update on this.

**Comment by flyswatter** [comment]
Good initial research.

Since Ledger has apps, I wonder if we could convince them to add message signing support.

In addition to removing buttons on the approval screen, I'd want a message, and then there is still the question of what the UI to pair should look like.

**Comment by izqui** [comment]
Quick update: I emailed Ledger support, and referenced me to their public trello (https://trello.com/b/5nQ1mdzt/ledger-roadmap) where there is no reference of this feature/bug. 

I was invited to join their Slack to continue the conversation.

Regarding confirmation screen, this is what they have in their Ethereum Wallet implementation (MEW doesn't show any UI related to the ledger)
<img width="542" alt="2017-01-30 at 4 45 59 pm" src="https://cloud.githubusercontent.com/assets/447328/22429689/b7e12a16-e70b-11e6-882f-08e339a72811.png">

**Comment by Recmo** [comment]
@LogvinovLeon and I implemented a `LedgerWalletProvider` that connects to the Ledger wallet over U2F (source)[1]. This should be an easier starting point than the MyEtherWallet code.

[1]: https://github.com/Neufund/ledger-wallet-provider

**Comment by Recmo** [comment]
The Ledger wallets are based on a HD wallets and use derivation paths. This is a bit of a mismatch with Ethereuem/web3 that is based on a flat list of accounts.

The Ledger supports (as far as I have tested) any derivation path of the form `m/44'/60'/*`where `*` can be empty. (`61` for ETC)

Their own Ethereum Wallet currently only uses the address available on `m/44'/60'/0'/0`. From their slack channel future versions will support multiple addresses, of the form `m/44'/60'/${account}'/${address}`. But this is not final and will be given further thought before implementation.

MyEtherWallet uses addresses available on `m/44'/60'/0'/${address}`.

For the `LedgerWalletProvider` we currently only implement `m/44'/60'/0'/0` (our goal is to be compatible rather than flexible).

**Comment by flyswatter** [comment]
@Recmo That is fantastic stuff! The linked code is a perfect starting point.

I'll need to either fork it to break out the core functionality or wrap it strangely though, since we don't use signing subproviders directly.

Maybe once I break that out into a small module, we could share that piece to share improvements and fixes.

Anyways, thanks a lot, I hope to work off this soon!

**Comment by Recmo** [comment]
Thanks! Feel free to fork it, we are very open to pull-requests. It implements the Web3-providers API, but has no dependencies on it. You should be able to use it on it's own just as easily:

```
const Ledger = require('ledger-wallet-provider')
ledger = new Ledger()
ledger.getAccounts(console.log)
ledger.signTransaction(...
```

The happy-paths are working, we will soon be adding more error handling and feedback (ie. "Browser unsupported").

**Comment by flyswatter** [comment]
Oh that surface is actually closer to what I wanted than I thought! Very cool!

**Comment by chafey** [comment]
+1, would love to see this feature done.  I have a ledger nano s and ledger blue so can help test, maybe code if given some guidance

**Comment by linagee** [comment]
@flyswatter I think the Ledger Nano S can get access to contract data. (Through an APDU.) Because it's able to look for function signatures to find tokens and display the proper Ethereum token name on the device. (Or at least in theory that's what their Github shows, haven't gotten that working yet on mine.)

**Comment by flyswatter** [comment]
Noting another nice looking ledger client available: https://github.com/LedgerHQ/ledger-node-js-api

**Comment by MicahZoltu** [comment]
Cross posting my [comment](https://github.com/ethereum/EIPs/issues/84#issuecomment-292402851) here.  _Please_ don't do what MEW and the Ledger app did (rumors suggest it was an accident that made it to release and now change is hard) and use a `m/44'/*` path without actually following the rest of the ([BIP 44 spec](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#path-levels)).  When you all get around to implementing this, please follow the spec and either create a new purpose (first segment) or use purpose `44'` and follow the spec associated with that : `m / purpose' / coin_type' / account' / change / address_index`

Alternatively, do something totally different with regards to the path, just don't have it start with `44'`.  :)

Soapbox aside, what are your plans for this?  As a dApp engineer working on Ledger integration, compatibility with MetaMask is important to us and we would like to make sure we are on the same page with the various wallet/integration providers.

You can see what other wallets are doing over in that thtread (I believe you all already have some representation there) but it is more even more important to follow the standard when integrating with a HW wallet because users will expect the hardware wallet to show the same accounts no matter which tool they use to interact with the wallet.  It is unfortunate that a bug in an early release of the Ledger app has resulted in this much community disconnect, but I think getting back on track with current standards until there is an official Ethereum one is critical for good UX for all of our users.

**Comment by mcjp78** [comment]
Do you have an ETA on this? It would really enhance the functionality of the ledger to use smart contracts.

**Comment by WestonReed** [comment]
I think this would be an excellent feature, and add a lot of security and usability to MetaMask. +1

**Comment by flyswatter** [comment]
For the people here asking for an update: At MetaMask, we also want this feature incredibly badly, and are currently doing some UX preparatory work to pave the way for a good integration.

Things have been a little slow to progress lately, we've had to put more time on fixes and support than usual, which has been a healthy expression of a still-exponential growth rate.

That said, after a few more core tuneups around how we manage nonces, and maybe a little token support, this is one of my top goals for MetaMask.

I think you can probably expect this by the end of the summer.

**Comment by woodydeck** [comment]
It's absolutely vital the hardware wallet support. There will be zero adoption of ethereum without commerce being made easy (metamask), and there will be no commerce unless there is a safe way to transact. Too many trojans around for people to ever trust anything more than walking around money in metamask.  I would prefer metamask to be my main hotwallet with a ledger or Trezor.

**Comment by burz** [comment]
Any updates on this? Ledger support would be very nice.

**Comment by Zanibas** [comment]
We've been burdened with meeting demands for token support and maintaining bug fixes that we have not begun work on hardware support, even though several members of the team really want to get started on it. We'd love it if the community could help.

**Comment by burz** [comment]
@Zanibas: Awesome. Is there any documentation by any chance on how one would have to interface with MetaMask and the hardware wallets (or I guess wallets in general) in order to get this done?

**Comment by kumavis** [comment]
@burz first part is to match the public API of our keyring class
https://github.com/metamask/eth-simple-keyring
https://github.com/metamask/eth-hd-keyring
then next step will be to do the ui elements

**Comment by flyswatter** [comment]
Nice looking tutorial on Ledger integrating code: https://blog.neufund.org/integrating-ledger-nanos-into-your-dapp-using-neufunds-js-wallet-provider-ecb71ca8f7b6

**Comment by vonpupp** [comment]
Support for ledger nano s would be awesome! It is becoming the most popular hw wallet.

**Comment by danfinlay** [comment]
Ledger reached out to let us know about a new integration guide:

https://blog.neufund.org/integrating-ledger-nanos-into-your-dapp-using-neufunds-js-wallet-provider-ecb71ca8f7b6

**Comment by LogvinovLeon** [comment]
@danfinlay Looks like it's the same guide You've posted on the 1st of September ;)

**Comment by danfinlay** [comment]
Whoops. Well, should get some time towards this a bit after Devcon.

**Comment by WyseNynja** [comment]
Any progress on this? Anything blocking that I can maybe help with?

**Comment by Zanibas** [comment]
@WyseNynja if you have any ideas on what an ideal interface for what the integration would look like, post them here! Also, see @kumavis last post above for details about the API.

The new excuse is: Cryptokitties and the resultant explosion of users has slowed us down again. Perhaps now is a great time for us to put this up to a bounty.

**Comment by drreen** [comment]
This!

**Comment by alexwagg** [comment]
Hello,
What is the status on this? What can I do to help? If you send some links I will look into things, this is a very very important feature for metamask, surprised you didn't have this already!

**Comment by danfinlay** [comment]
Partly, we still need design on this. Our one designer is stretched pretty thin lately, we have a full new UI coming very soon. Once it's launched, we will be putting some focus on new account types, including hardware.

Until that general protocol is established, this is a very transitional time to try to contribute to this feature.

My goal is to make "new signers" easy to add, and this requires a bit more consideration than just cramming in Ledger or Trezor support. We're trying to do it right, and pave the way for a continuously good user experience.

Also, this week, we're completely consumed in gas related issues, scaling issues, issues that are hit by the many new users who keep coming on to the platform.

**Comment by gre** [comment]
I was about to start working on this and per @danfinlay comment, I'm going to HODL a bit (after investigated a bit, I figured out what just need to be done is to implement a "Keyring" class, and this should be quite straightforward. but on the UI perspective, obviously there are things to figure out).

This feature means that MetaMask could basically be a proxy to the Ledger device and that would be HUGE because you benefit both of the hardware wallet security and you benefit that MetaMask is supported on many websites.

Anyway, feel free to ping me if you need help integrating things, we have recently refactored our JS libraries @LedgerHQ ( https://github.com/LedgerHQ/ledger-node-js-api/pull/37 which should be merged soon)

**Comment by bluebirdly** [comment]
@Recmo is getMultipleAccounts(path, offset, accountno) operational please? I've tried the standard path = "44'/60'/0'/0", but the callback return empty.

**Comment by Rafaqut** [comment]
Can't wait for this to be implemented. I really like using metamask!

**Comment by chuckishungry** [comment]
It would be great if MetaMask support for Ledger devices was prioritized. Lots of people are switching to MEW because they already support Ledger Nano and Nano S.

**Comment by cslarson** [comment]
You can't use MEW in with a dapp though so yeah for simple transfers that's fine, but to use something like create a [dai cdp](https://dai.makerdao.com/) MEW doesn't help and having hardware wallet support would be wonderful.

Just throwing the idea out there @danfinlay but is there a bounty the community (r/ethtrader, for example) could contribute to to expedite hardware wallet support in metamask?

**Comment by benji7425** [comment]
@cslarson you actually can use MEW for DAI CDP if you're willing to fiddle around with the contract ABI in MEW. It's not particularly easy but I have done it before when wanting to store a CDP on my ledger.

I found the best way is to transfer ETH out to MM to create the CDP, then 'give' the CDP back to my ledger address. Then the only thing you need to use MEW to do is call the 'give' function on the contract to return it to your MM address to manage it again.

**Comment by benji7425** [comment]
Following on from the bounty suggestion, https://gitcoin.co could be a good option

**Comment by gre** [comment]
following previous comment, we have released at Ledger (and continue improving on) a starter kit to make dapps with both metamask and ledger device: https://www.npmjs.com/package/create-dapp . **NB: In that starter, the current auth workflow is basically something similar to MEW**: which probably isn't the final answer to the problem exposed in here (because dapps need to manually migrate to this approach), but at least it's a working solution we have today.

That said, it's totally possible to have metamask directly connecting to the Ledger device, and, actually, you can inspire from our work implementing this starter kit. basically you can use [@ledgerhq/web3-subprovider](https://github.com/LedgerHQ/ledgerjs/tree/master/packages/web3-subprovider) library that implements a subprovider to work with provider-engine. This library is still bleeding edge and API is likely to change a bit.

**Comment by danfinlay** [comment]
Again, this is a top priority for us at MetaMask, but we're not going to just throw it together, we're ensuring that the way we add multiple signing strategies is modular, so that we can more easily add different signers in the future.

We are mostly bottlenecked by our number of core contributors, and we are actively hiring to relieve this bottleneck.

While we are big fans of bounties, and have used GitCoin for a variety of enhancements and bug fixes already, we are unlikely to use bounties for this particular feature, since it's highly architectural, and involves changes with a large portion of the project, so it would require tight enough integration with the core team that it might as well be done by the core team anyways.

I promise, this is one of our top priorities for the moment, up there with per-domain log-in for privacy. No bounty nor `+1` on this thread is going to speed it up. It's already a top priority. We are simply scaling our currently very small team while balancing maintenance, integrations, and support.

**Comment by mickayz** [comment]
One thought I have had is that maybe Metamask can be forked and have all the guts of transaction signing yanked out and replaced with the ledger API?  That way the forked Metamask wont have to worry about multiple accounts, just the one with the ledger.

Of course, that wouldn't be worth doing if Metamask will be adding this feature immediately after, but if it will be taking some time to support, then maybe it is worthwhile as a stop gap solution? 

Right now users that choose to protect their funds in the best way possible are being excluded from much of the blooming dApp ecosystem, which is a pretty big bummer.

**Comment by lazaridiscom** [comment]
>No bounty [...] on this thread is going to speed it up.

@danfinlay , dare to bet?

**Comment by danfinlay** [comment]
> One thought I have had is that maybe Metamask can be forked and have all the guts of transaction signing yanked out and replaced with the ledger API? That way the forked Metamask wont have to worry about multiple accounts, just the one with the ledger.

Be my guest. Everyone says they want to fork whenever we're slow on a feature, but actually maintaining software is hard.

> @danfinlay , dare to bet?

I mean, if you're going to put it that way, I'd just have to clarify: We would hold this feature to the highest standard, and be very picky about what we had accepted.

That said, what size bounty would get you pursuing it on the side? You've done good bounty work so far. I might be willing to write up acceptance criteria for those brash enough to dare the challenge.

**Comment by mickayz** [comment]
To clarify, I meant no disrespect.  I appreciate the work you and your team put into this project and understand that you don't owe us anything.  I only mean to brainstorm solutions for the impatient in case anyone has the resources to create an intermediate solution. Thank you for supporting this project :)

**Comment by danfinlay** [comment]
Forking is a bit like war: we all know it's on the table.

**Comment by lazaridiscom** [comment]
>We would hold this feature to the highest standard, and be very picky about what we had accepted.

We agree here.

> That said, what size bounty would get you pursuing it on the side?

No idea at this point, really. But seems to be an x.0 ETH bounty. And sometimes, working on an issues triggers a cascade of new issues (had this before). But one has not to set the bounty price in advance. Can be done on "fair-play" / "tip" basis, too

> I might be willing to write up acceptance criteria for those brash enough to dare the challenge.

Possibly the way to go is to place a pre-bounty. I mean, specifying the requirements and acceptance-criteria is a task in itself. And work on specs/test/acceptance criteria or even instructions for other developers sometimes pops up new ways of solutions.

In any way, if you like, just provide a very draft of acceptance criteria, kind of bullet list. Will have this one on my radar.

**Comment by danfinlay** [comment]
Right now my apprehension about the bounty is that GitCoin lets people "lock" an issue that they have "checked out". This one would absolutely be posted on a basis of "you can race us if you want", because I don't plan on waiting for anyone to do this, I just happen to be way busier than I want to be.

In any case, I wrote up a pretty decent doc on my current thinking for this issue here:
https://github.com/MetaMask/metamask-extension/issues/3249

**Comment by lazaridiscom** [comment]
@danfinlay , sounds rational. "wild-west" style bounty, I guess.

Wow! You wrote this beast (#3249)  just now?

I'm preparing to "dive" into code-level for another issue (dive = isolation for 78 hours). Hopefully I succeed and will be then mentally free to take a look on all your stuff here.

.

**Comment by danfinlay** [comment]
> Wow! You wrote this beast (#3249) just now?

I've been thinking about it for a while. ;)

Happy to chat about it any time, as you'll see there, it's a legitimately hard problem, with a lot of important considerations (to get it very right).

**Comment by koalalorenzo** [comment]
Anybody still working on this?

**Comment by marceljay** [comment]
I'm interested to. I'd like to securely invest in ERC-Token Projects and I'm usually doing that through IDEX or Forkdelta(Ed) using metamask.

**Comment by danfinlay** [comment]
> Anybody still working on this?

Yes. Currently lazaridiscom is doing research towards a standard architecture that will allow us to add new signers much more flexibly in the future.

We also some other devs gearing up to start working on this soon, and they will hopefully be able to benefit from his research.

We are not rushing out this feature, we are trying to do it right. Please [read above](https://github.com/MetaMask/metamask-extension/issues/717#issuecomment-365416522) if you have any other questions. We are actively working on this.

**Comment by chuckishungry** [comment]
Thanks @danfinlay 

To be clear, there is still no timeline for release of Ledger support. Is that correct?

**Comment by danfinlay** [comment]
@chuckishungry I'm not completely clear what you want here. How's this?:

- Research phase estimated at 30-60 days.
- Development phase estimated at 30-60 days.

Estimated time range: 60-120 days.

**Comment by jamespic** [comment]
I wanted to help out, so following @kumavis suggestion, I've started work on an implementation of the Keyring class: https://github.com/jamespic/eth-ledger-keyring. It's still very rough, but figured I'd put it up here to get involved in the process.

**Comment by jamespic** [comment]
One thing that I wasn't sure about is how to handle devices that aren't always there, or multiple devices, since I couldn't see anything in the keyring interface that speaks to this. I suspect this might relate to the UI design stuff in #3249, so I thought it would be useful to call it out.

**Comment by lazaridiscom** [comment]
@jamespic , very nice! I don't want to look at the code at this point. Can you quickly describe what functionality already works, and most importantly, if you have you tested this with a real ledger yet?

> I wasn't sure about is how to handle devices that aren't always there, or multiple devices

you mean ledger-devices?

**Comment by jamespic** [comment]
@lazaridiscom So far it can:
- Create and list accounts (using the same BIP39 derivations as other wallets that support Ledger)
- Sign transactions
- Sign personal messages

It can't sign arbitrary hashes of typed message, or export private keys, because the devices do not support these features.

I've got tests that test these features. I've tested them with a real device, and I've used Ledger's `hw-transport-mocker` library, which allowed me to run the tests once against a real device, record the interactions, and use the recording as an oracle so future tests can be run without a physical device.

What I haven't tested, is any integration with the MetaMask extension. I'll probably try this out soon, but it'll mean building jury-rigged versions of `KeyringController` and `metamask-extension` to try it out.

In terms of "devices that aren't always there, or multiple devices", I'm mostly thinking of Ledger devices now, but the same issue is likely to occur with other hardware devices. The current code keeps a list of addresses it expects to find on the attached Ledger device, but what if someone plugs in a different device, with a different set of addresses? Or there's no device plugged in? At the moment, it just fails loudly if you try to sign something but the keys don't match, but that's probably not the UX we want.

**Comment by lazaridiscom** [comment]
Sounds very(!) good.

"Device" stuff: ok, I understand now. This sounds like "doable detail work", so all fine.

Re integration: within #3427, I am not yet ready to go for integration, need a few more days for sure, possibly a week. But will do this anyways (instead of a dummy, I use your module). So possibly it would be better if you focus to "polish" your module a bit more.

@danfinlay , not sure here, but you should verify if @jamespic could "Start Work" on the open bounty within #3249, with the deliverable changed to:

* production grade `eth-ledger-keyring` module, which lives within a MetaMask organization repository.
 
(production grade: standard-js-documentation (JsDoc i guess? ), unit/integrations tests provided/or support the test-writer etc.)

**Comment by jamespic** [comment]
Don't worry about adding me to the bounty. I'm doing this solely so that I can use my Ledger from MetaMask, and I think I'll get that faster if I'm not diluting any bounty.

**Comment by lazaridiscom** [comment]
@jamespic , my understanding is that the MetaMask project is open to any form of contributions. I made the suggestion for fair-play reasons.

> and I think I'll get that faster if I'm not diluting any bounty.

Nah, bounties make things usually faster. 

But ok, so I ask you just for a bit help in a usual form:

Please try to add some documentation (even without a standard-format like JsDoc). I would need documentation about "what is different to the standard keyring", and possibly the usual "strange things docu". As said, i expect to use your result in the next days.

And then I guess I need to order a ledger one week earlier than planned.

In the meantime, if you like to invest time "building jury-rigged versions" to try this out by yourself, its of course ok so. Just please share your findings if you like.

**Comment by MicahZoltu** [comment]
@jamespic I recommend against using the LedgerJS library directly, unless it has undergone some major improvements.  It is way too easy to lock up the ledger and/or the app since the USB connection is raw and the protocol doesn't guarantee message separation and the ledger gets very unhappy when it receives multiple messages in a row without responding to the first.

I wrote a library a while back to deal with this and hide most of the complexity from the user: https://github.com/ethereumjs/ethereumjs-ledger

It is possible they have improved their JS libraries so they no longer have the problems, so certainly do a bit of research and don't take my word for it.  😄 The easiest way I found to break a ledger is to simply fire a bunch of requests in a row to the ledger without waiting for them to complete.

**Comment by MicahZoltu** [comment]
Just took a quick glance at their new libraries and it appears they still haven't addressed the major issue which is locking the connection to prevent (minimize) the ledger getting two messages in a row without responding to the first.

**Comment by jamespic** [comment]
Thanks for the heads up @MicahZoltu. I'm on the fence about whether to use your library, or just borrow your locking solution - the newer Ledger libraries support `signPersonalMessage`, which it looks like `ethereumjs-ledger` would need updating to support. I'll have a play with the code, and see how it goes.

**Comment by MicahZoltu** [comment]
Feel free to submit a PR, fork, or simply crib the code in that library.

**Comment by jamespic** [comment]
@lazaridiscom Yes, documenting the code is next on my agenda.

You're right that I don't expect the multi-device stuff to be difficult. Once we know what our approach is, it should be an easy refactor. The question is what the UI's expectations of the keyring are. I can see two obvious choices (but there may also be others):
- The controller creates one keyring instance, that keeps track of the ledger devices it knows about. This keyring won't always return the same list of addresses.
- The controller creates one keyring instance per device. These keyrings always return the same list of addresses, but should have a way of signaling whether the device they control is available.

Both options are easy enough, but I need to choose between them.

**Comment by lazaridiscom** [comment]
@MicahZoltu , sounds like a solid fundament.

@jamespic , reminding you "I am not yet ready to go for integration, need a few more days for sure, possibly a week.", so me out at this point.

@all, naturally, you are all free to continue on this right away.

**Comment by jamespic** [comment]
@lazaridiscom Fair enough. Let's leave the question of how to integrate until you're ready. If we need to do something different,  the changes should be minor and localised.

**Comment by danfinlay** [comment]
@jamespic A third option that I would endorse:

- The controller creates one keyring per derivation path per device. This way we serialize the derivation path along with the device ID and the accounts we expect, which is everything we should require to recognize a device and request a valid signature.

Your second option is otherwise my favorite, but it's just an opinion, I agree either could work.

The visual aspect was what I got to before being swept up with other things last year. I think that signers should usually get their own "signing UI", at least so they can explain "please connect the device, please enter your PIN", or whatever custom steps that device has.

I think these screens would probably appear after the regular tx confirmation screen. Ideally they could show custom views, like an animation of connecting the Ledger, but for an MVP, I'd be okay if we just made a view that could display relevant text from the signing keyring while a signature is pending. Again, curious to see what other MVP you come up with. I'm happy if we/others can just fork your work to embellish the UI, if that's how it works out, so thanks for taking this on!

I'd be happy to pay out the bounty if this was delivered in a shippable form, but if James isn't interested in a bounty, I'm not going to force it on him.

**Comment by jamespic** [comment]
@danfinlay Option 3 sounds sensible, and I think is actually less refactoring than the other two. It's more-or-less how the code works now, I'd just need to add some checks on the currently connected device.

When I've had a look at the UI code for MetaMask in the past, I've struggled to grok it. I'll take another look, but I'm very happy for someone who already knows the UI code to fork what I've done so far and run with it.

**Comment by danfinlay** [comment]
I’m happy to run with your work later, or give you a short video tour, since this is a high impact contribution.

**Comment by jamespic** [comment]
There's probably no need to a walkthrough - I've had another look through the code, and I reckon I get it. Am I right in guessing that new features are only targetting the new UI?

**Comment by danfinlay** [comment]
That’s right, New UI only for new features.

**Comment by jamespic** [comment]
I've got some very crude WIP code at https://github.com/jamespic/metamask-extension, but I've hit a brick wall. Ledger devices can be accessed as either HID or U2F devices, but AFAICT, Chrome doesn't expose either interface to extensions. The HID API only seems to be accessible to chrome apps, and U2F doesn't seem to be working either, due to alleged origin policy violations (I *think* this is because U2F will only work if the protocol is `https`, and in the case of Chrome extensions it's `chrome-extension` although that seems a bit mad, so I'll investigate further)

**Comment by MicahZoltu** [comment]
This actually aligns with some of my prior experiences.  I don't believe you will find an easy workaround for this problem.  You may want to open a Chrome bug/feature request.

**Comment by danfinlay** [comment]
That's rough, @jamespic, and makes me think even more that maybe we need to provide a desktop client for managing the accounts.

**Comment by gre** [comment]
@jamespic yes, https is required for U2F. webusb requires https as well, so we have no other choices at the moment. I think crypto websites should use HTTPS anyway right?
We might want to disable the "ledger support" feature if the page is not on HTTPS / show a message.
So you said however that U2F is not exposed on the extension side? could you check if WebUSB is? WebUSB is the future and should eventually land as a replacement for u2f for the ledger devices.

**Comment by danfinlay** [comment]
> crypto websites should use HTTPS anyway right?

This is a browser extension. We're talking about letting the browser extension request signatures from the Ledger device. Extensions don't get `https`, they get the `chrome://` prefix.

**Comment by gre** [comment]
But you can execute code on the background script that runs on the page no?

**Comment by gre** [comment]
Oh yeah, so U2F also have this origin policy thing, so I tend to think @jamespic issue is related to that. Extensions should be considered secure so I guess WebUSB will work in them. There is this ongoing discussion but it's unclear if WebUSB is supported in extension https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/kRd0lD6VwtA (& not so easy to find documentation on the subject)

**Comment by jamespic** [comment]
It looks like it's an easy fix to make this work in Chrome (https://github.com/jamespic/chromium/commit/69ed27d5d03838a00a6dfe10d36b7c0f19346578 - I tested it, it seemed to work), however the Chromium devs have already come back and said WontFix (https://bugs.chromium.org/p/chromium/issues/detail?id=823736).

That pushes us towards WebUSB, which I was hoping to avoid, since (a) I've been burned by draft W3C standards before, and (b) it'd be way more work, because there's no existing WebUSB Ledger code I can re-use. However it does appear to be available to extensions (based on a very limited sanity check I just did), and might give a better UX (tunnelling Ledger stuff through U2F is an ugly hack, and it shows).

**Comment by jamespic** [comment]
The WebUSB support in extensions doesn't actually look usable. It's got the APIs, but always fails to find any devices, whereas the Ledger shows up when I call them from a regular browser tab.

**Comment by gre** [comment]
@jamespic you can try this snippet in the devtool console, it should work if you have a ledger device connected & unlocked

```js
const ledgerDevices = [
 { vendorId: 0x2581, productId: 0x3b7c },
 { vendorId: 0x2c97 }
]
await navigator.usb.requestDevice({ filters: ledgerDevices })
devices = await navigator.usb.getDevices()
console.log(devices)
devices[0].open()
```

**NB:** however you won't be able to communicate yet with the device as we don't expose an interface from webusb yet but we're experimenting on it these days. You can see latest experiment [here](https://github.com/gre/ledgerjs/blob/experimental/webusb/packages/hw-transport-webusb/src/TransportWebUSB.js) **but it does not work with the current Eth ledger app**, hopefully if we can figure it out, we will add support to it soon :)

**Comment by jamespic** [comment]
@gre Trying a snippet much like that one was what persuaded me that WebUSB doesn't work from extensions, but maybe there's something I've missed, so I'll look again

**Comment by gre** [comment]
@jamespic oh ok. maybe WebUSB in context of an extensions is different. API is here but not seeing the device?

and does `requestDevice` popup open with the permission UI ? on a normal webpage:

<img width="496" alt="capture d ecran 2018-03-21 a 10 43 55" src="https://user-images.githubusercontent.com/211411/37703144-fafa75b6-2cf4-11e8-8470-7e78f2e6c136.png">

**Comment by jamespic** [comment]
@gre Yes, if I do that on a normal page, I get the same thing you do. However if I open up a developer console for one once the MetaMask pages (the background page is where all the signing happens, but any page will do), `requestDevice` throws an exception without displaying the device selection screen.

**Comment by lazaridiscom** [comment]
>There is currently no clean way of determining an appropriate origin of an extension for the purposes of populating clientData in the U2F protocol

@jamespic , looking quickly at all this, it seems like solving the above (finding a clean way, so the ["wontfix"](https://bugs.chromium.org/p/chromium/issues/detail?id=823736) is revised) would solve this within chrome, and thus the interfacing problem?

Does it work with e.g firefox?

**Comment by jamespic** [comment]
@lazaridiscom It's conceivable that Chromium devs would be satisfied with some sort of solution where extensions can effectively act like any origin, since they already do this with the `fetch` API in extensions. Although even a simple fix would be a code change to Chrome.

I'll check what Firefox does with U2F and WebUSB requests from extensions when I get chance.

**Comment by jamespic** [comment]
Firefox doesn't seem to have working U2F in extensions either - the `window.u2f` object is there, unlike Chrome, but requests always fail, possibly due to the domain being ill-defined, as the Chromium devs alluded to.

I can't seem to get WebUSB working in Firefox at all, so that doesn't look promising either.

**Comment by lazaridiscom** [comment]
Ledger arrived, follow-up: #4060

**Comment by lazaridiscom** [comment]
@gre, is there any way to set the ledger in a kind of development-mode, thus one does not have to unlock it again and again?

**Comment by lazaridiscom** [comment]
@gre, @jamespic - can you please verify the new solution paths mentioned within #4204 (extension-to-extension communication, extensions-messaging)?

@danfinlay the "heavy bounties" mentioned above would go exclusively to the #4205 issue. A "high-profile" bounty (I guess minimum 5 ETH, if not 10) could generate enough visibility and interest, thus a chrome dev is attracted to a) find a solution and b) ensures that it is finally merged (which can be the most difficult thing).

**Comment by danfinlay** [comment]
I'll look into a heavy bounty for that task, in hopes of getting a Chrome dev's interest.

**Comment by lazaridiscom** [comment]
Bounty for #4205 makes now even more sense (WebUSB path, chrome devs).

**Comment by lazaridiscom** [comment]
> @gre, is there any way to set the ledger in a kind of development-mode, thus one does not have to unlock it again and again?

Answering myself: its a standard-option in the ledger-device menu.

**Comment by lazaridiscom** [comment]
superseded by #4060

(similar/related: #4625, Trezor)

( @danfinlay , not 100% sure here, but this should be possibly close, too)

**Comment by lazaridiscom** [comment]
@ all

you can help by starring a relevant issue:

https://bugs.chromium.org/p/chromium/issues/detail?id=770896

**Comment by matkam** [comment]
Now that Firefox has U2F support in Firefox 67, can you enable Ledger support?

**Comment by matkam** [comment]
Made a pull request to enable Leger wallets in Firefox https://github.com/MetaMask/metamask-extension/pull/6659
